cpp/src/io/avro/avro.cpp

```diff
@@ -96,19 +96,21 @@ bool container::parse(file_metadata* md, size_t max_num_rows, size_t first_row)
   md->sync_marker[1] = get_raw<uint64_t>();
 
   md->metadata_size  = m_cur - m_base;
-  md->skip_rows      = 0;
+  md->skip_rows      = first_row;
+  md->total_num_rows = 0;
   max_block_size     = 0;
   total_object_count = 0;
   while (m_cur + 18 < m_end && total_object_count < max_num_rows) {
     auto const object_count = static_cast<uint32_t>(get_encoded<int64_t>());
     auto const block_size   = static_cast<uint32_t>(get_encoded<int64_t>());
     if (block_size <= 0 || object_count <= 0 || m_cur + block_size + 16 > m_end) { break; }
+    md->total_num_rows += object_count;
     if (object_count > first_row) {
       auto block_row = static_cast<uint32_t>(total_object_count);
       max_block_size = std::max(max_block_size, block_size);
       total_object_count += object_count;
       if (!md->block_list.size()) {
-        md->skip_rows = static_cast<uint32_t>(first_row);
+        md->skip_rows = first_row;
```

Comment:
This was assigned in line 99. So maybe line 99 or this line is incorrect.

Reply:
Yeah the initial `md->skip_rows = first_row;` assignment that I added was key to actually solving the bug of skip rows being ignored.

And this subsequent assignment you've highlighted should be redundant, as it's within the `if (!md->block_list.size())` scope, so it should only be hit on the first block.

i.e., this should hold:

```diff
-        md->skip_rows = first_row;
+        CUDF_EXPECTS(md->skip_rows == first_row, "Invariant check failed: md->skip_rows != first_row");
```

However, a few lines up, the whole thing is guarded by `if (object_count > first_row) `... then `else { first_row -= object_count }`.

That seems problematic.  What if the first block has less than or equal rows to the requested number of rows to be skipped? (i.e. the value of `first_rows` at that comparison).  Also, in that case, the block doesn't even get added to the `block_list`, which doesn't seem correct.

I'm going to review a bit more.  At the very least I think there needs to be some tests that exercise the multiple block logic.

Reply:
(The suggested code isn't correct, but I think the analysis is on point.)

Reply:
I did some digging and thankfully `fastavro` has support for writing multiple blocks via a `sync_interval=` param to the writer.  I'm going to write a few tests to exercise this logic now.  Then we can reconvene and review the proposed C++ changes.